### DISCUSSION TITLE
Best way to access Fastify instance from controllers in a separate file

### DISCUSSION THREAD

**Discussion by StanleySathler** [fastify/fastify#3770] [General] [answered]
Hey guys 👋 
I've been playing around with Fastify after some time working with Express.

I see we can add decorators to add custom values to Request, Reply and Fastify global instance (aka. Root Context?).

Most tutorials on Web shows handlers (aka. controllers?) in the same file as the routers, which makes much easier to access the root Fastify instance to access the decorated values.

However, usually, controllers go in a different file. This means they don't have easy access to root Fastify instance unless:
1. The handlers are thunks that accept Fastify instance as a param;
2. We make a Singleton so that we can import Fastify instance from the controllers.

I'm wondering what's the best way to access the Fastify instance across different files. 

The same question was made on StackOverflow once, but I'm wondering if those answers align with best practices recommended by Fastify team:
https://stackoverflow.com/questions/52939927/how-to-access-fastify-instance-from-a-handler-controller-file

Thanks in advance! 🚀

**Comment by zekth** [comment, marked as answer]
You have a reference to the fastify instance inside the request. eg:
```js
handler: (request,reply) => { request.server }
```

This has been added in : https://github.com/fastify/fastify/pull/3102

See the type definition: https://github.com/fastify/fastify/blob/06a96cb7c67d6605f5b22de8e4784a24bdc6469b/types/request.d.ts#L35

**Reply by StanleySathler** [reply]
Amazing, @zekth! That's exactly what I needed. Thank you! 🚀

**Comment by Eomm** [comment]
There is another option: the `this` context but you must use a named function:

```
handler: function (request, reply) { 
  assert(this === request.server)
}
```

**Reply by StanleySathler** [reply]
Good to know we have this option as well @Eomm 🙌 Thanks!
